docs: auth passwords page (#19490)
supabase/supabase: apps/docs/pages/guides/auth/passwords.mdx

import Layout from '~/layouts/DefaultGuideLayout'
import { Accordion } from 'ui'

export const meta = {
  id: 'auth-passwords',
  title: 'Passwords',
  description: 'How to work with passwords in Supabase Auth',
}

> Using passwords to authenticate users is a tried-and-tested method to give your users access to your application. Supabase Auth provides you with secure configuration options and uses best-practices to store and verify your user's passwords.

Reviewer: ```suggestion
Using passwords to authenticate users is a tried-and-tested method to give your users access to your application. Supabase Auth provides you with secure configuration options and uses best practices to store and verify your user's passwords.
```

## Password security

> Passwords derive their security from how easy they are to guess or brute-force In theory this means that the longer a password is and the more required characters it has (digits, lowercase and uppercase letters, symbols) the harder it will be to guess.

Reviewer: ```suggestion
A password is more secure if it is harder to guess or brute-force. In theory, a password is harder to guess if it is longer. It is also harder to guess if it uses a larger set of characters (for example, digits, lowercase and uppercase letters, and symbols).
```

This table shows the _minimum_ number of guesses that need to be tried to access a user's account:

| Required characters                          | Length | Guesses          |
| -------------------------------------------- | ------ | ---------------- |
| Digits only                                  | 8      | ~ 2<sup>27</sup> |
| Digits and letters                           | 8      | ~ 2<sup>39</sup> |
| Digits, lower and uppercase letters          | 8      | ~ 2<sup>48</sup> |

> | Digits, lower and uppercase letters, symbols | 8      | ~ 2<sup>52</sup> |

Author: Basically `log2((characters)^(length))`...

> In reality though, passwords are not always generated at random, but often contain variations of names, words, dates, sentences or phrases. Malicious actors can use these properties to guess a password in less attempts.

Reviewer: ```suggestion
In reality though, passwords are not always generated at random. They often contain variations of names, words, dates, and common phrases. Malicious actors can use these properties to guess a password in fewer attempts.
```

> There are hundreds of millions (and growing!) known passwords out there. Malicious actors can use these lists of leaked passwords to automate login attempts (known as credential stuffing) and steal or access sensitive user data.

Reviewer: Also, passwords are often reused. Malicious actors can use lists of leaked passwords to automate login attempts (known as credential stuffing).

Author: Would actually prefer to bring up the leaked passwords first.

### Password strength and leaked password protection

> To help protect your users data and accounts, Supabase Auth allows you fine-grained control over the strength of the passwords used on your project. You can configure these in your project's [Auth settings](/dashboard/project/_/settings/auth):

Reviewer: ```suggestion
To help protect your users, Supabase Auth allows you fine-grained control over the strength of the passwords used on your project. You can configure these in your project's [Auth settings](/dashboard/project/_/settings/auth):
```

- Set a large minimum password length. Anything less than 8 characters is not recommended.

> - Set the required characters that must appear at least once in a user's password to be accepted. Use the strongest option of requiring digits, lowercase and uppercase letters and symbols.

Reviewer: ```suggestion
- Set the required characters that must appear at least once in a user's password. Use the strongest option of requiring digits, lowercase and uppercase letters, and symbols.
```

- Prevent the use of leaked passwords. Supabase Auth uses the open-source [HaveIBeenPwned.org Pwned Passwords API](https://haveibeenpwned.com/Passwords) to reject passwords that have been leaked and are known by malicious actors.

### Additional recommendations

> In addition to choosing suitable password strength settings and preventing the use of leaked passwords, consider prominently asking your users to:

Reviewer: ```suggestion
In addition to choosing suitable password strength settings and preventing the use of leaked passwords, consider asking your users to:
```

- Use a password manager to store and generate passwords.
- Avoid password reuse across websites and apps.
- Avoid using personal information in passwords.

> - Use [Multi-Factor Authentication](/docs/guides/auth/auth-mfa) whenever possible.

Reviewer: ```suggestion
- Use [Multi-Factor Authentication](/docs/guides/auth/auth-mfa).
```

## Resetting a user's password (forgot password)

Strong passwords are difficult to remember, so Supabase Auth provides you with APIs to build a secure password reset flow.

### Overview

> To add a secure password reset flow in your application, follow these steps.

Reviewer: ```suggestion
To add a secure password reset flow to your application:
```

1. Build the **password reset page**:
    - Should be publicly accessible and contain a form asking for the user's email address.
    - Use the [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) API to request a password reset link for a user's email address.
    - Specify the `redirectTo` parameter when calling this API to point to the URL of the **change password page.**
2. Build the **password change page**:
    - Should be accessible only to authenticated users.
    - Add its URL to the allowed [Redirect URLs](/dashboard/project/_/auth/url-configuration) settings.

>     - Show a form or other propmt asking the user to choose a new password.

Reviewer: ```suggestion
    - Show a form or other prompt asking the user to choose a new password.
```

    - Call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) API to set a new password for the user.

> The email link sent to your users works very similarly to [passwordless authentication](/docs/guides/auth/passwordless-login):

Reviewer: ```suggestion
The email link sent to your users works in the same way as [passwordless authentication](/docs/guides/auth/passwordless-login):
```

> 1. After a user visits the **reset password page** they are sent a reset password link.

Reviewer: ```suggestion
1. After a user visits the **reset password page**, they are sent a reset password link.
```

>     - If you use PKCE (default) this link will be bound to the browser or device the user asked for password reset. Consider informing the user in the email message to use the same device or browser they requested the link from.

Reviewer: ```suggestion
    - If you use PKCE (default), this link only works on the device or browser where the original reset request was made. Display a message to the user to make sure they don't change devices or browsers.
```

>     - If you use the implicit grant flow, the link will be valid when opened on any device.

Reviewer: ```suggestion
    - If you use the implicit grant flow, the link can be opened on any device.
```

> 2. A user clicks the link and is taken first to Supabase Auth which validates it.

Reviewer: ```suggestion
2. A user clicks the link. They are taken to Supabase Auth, which validates the link.
```

>     - If the link is valid, it will redirect to the **change password page** which was specified in the `redirectTo` parameter.

Reviewer: ```suggestion
    - If the link is valid, it redirects to the **change password page**, which you specified in the `redirectTo` parameter.
```

>     - If the **change password page's** URL is not properly registered in the [Redirect URLs](/dashboard/project/_/auth/url-configuration) configuration, the user will be taken to the default Site URL page.

Reviewer: ```suggestion
    - If the **change password page's** URL is not properly registered in the [Redirect URLs](/dashboard/project/_/auth/url-configuration) configuration, the user is taken to the default Site URL page.
```

3. Supabase Auth redirects to the **change password page** including [session](/docs/guides/auth/sessions) information in the URL.

>     - If you used PKCE (default) when the user asked for a reset password link, the redirect will contain the `code` query param.

Reviewer: ```suggestion
    - If you used PKCE (default), the redirect contains the `code` query param.
```

    - If you are not using official Supabase libraries, or have a custom setup, extract the `code` from the URL and call the [`supabase.auth.exchangeCodeForSession`](/docs/reference/javascript/auth-exchangecodeforsession) API.

>     - In most cases using official Supabase libraries will handle this for you.

Reviewer: ```suggestion
    - Official Supabase libraries handle code the code exchange for you.
```

>     - If you used the implicit flow when the user asked for a reset password link, the redirect will contain a URL fragment encoding the user's [session](/docs/guides/auth/sessions).

Reviewer: ```suggestion
    - If you used the implicit flow, the redirect contains a URL fragment encoding the user's [session](/docs/guides/auth/sessions).
```

### Example: Request a password reset email

<Tabs
  scrollable
  size="small"
  type="underlined"
  defaultActiveId="js"
  queryGroup="language"
>
<TabPanel id="js" label="JavaScript">

> Supabase provides a convenient method [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectTo` which we will use to pass an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration) or it won't redirect the user.

Reviewer: ```suggestion
Supabase provides a convenient method [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectTo`, which is an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration).
```

```js
await supabase.auth.resetPasswordForEmail('hello@example.com', {
  redirectTo: 'http://example.com/account/update-password',
})
```

</TabPanel>
<TabPanel id="kotlin" label="Kotlin">

> Supabase provides a convenient method [`.sendRecoveryEmail`](/docs/reference/kotlin/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectUrl` which we will use to pass an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration) or it won't redirect the user.

Reviewer: ```suggestion
Supabase provides a convenient method [`.sendRecoveryEmail`](/docs/reference/kotlin/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectUrl`, which is an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration).
```

```kotlin
supabase.gotrue.sendRecoveryEmail(
    email = "hello@example.com",
    redirectUrl = "http://example.com/account/update-password"
)
```

> If you are on one of the Kotlin targets that have built-in support for redirect url handling like Android, you may want to checkout [OAuth and OTP link verification](https://supabase.com/docs/reference/kotlin/initializing).

Reviewer: ```suggestion
If you are on one of the Kotlin targets that have built-in support for redirect URL handling, such as Android, see [OAuth and OTP link verification](https://supabase.com/docs/reference/kotlin/initializing).
```

</TabPanel>
</Tabs>

### Example: Updating a user's password

<Tabs
  scrollable
  size="small"
  type="underlined"
  defaultActiveId="js"
  queryGroup="language"
>
<TabPanel id="js" label="JavaScript">

> To update the password we call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) method and pass along the new password to this method.

Reviewer: ```suggestion
To update the password, call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) method with the new password.
```

```js
await supabase.auth.updateUser({ password: new_password })
```

</TabPanel>
<TabPanel id="kotlin" label="Kotlin">

> To update the password we call the [`.modifyUser`](/docs/reference/kotlin/auth-updateuser) method and pass along the new password to this method.

Reviewer: ```suggestion
To update the password, call the [`.modifyUser`](/docs/reference/kotlin/auth-updateuser) method with the new password.
```

```kotlin
supabase.gotrue.modifyUser {
    password = "new_password"
}
```

</TabPanel>
</Tabs>

## Frequently asked questions

### How are passwords stored?

> Supabase Auth uses [bcrypt](https://en.wikipedia.org/wiki/Bcrypt), a strong password hashing function to store a hash of user's passwords. You would not be able to impersonate a user by having access to this information. Each hash is accompanied by a randomly generated salt parameter that provides an additional layer of protection.

Reviewer: ```suggestion
Supabase Auth uses [bcrypt](https://en.wikipedia.org/wiki/Bcrypt), a strong password hashing function, to store hashes of users' passwords. Only hashed passwords are stored. You cannot impersonate a user with the password hash. Each hash is accompanied by a randomly generated salt parameter for extra security.
```

> This information is stored in the `encrypted_password` column of the `auth.users` table. The column's name is a misnomer (cryptographic hashing is not encryption), but is kept for historical and backward-compatibility reasons.

Reviewer: ```suggestion
The hash is stored in the `encrypted_password` column of the `auth.users` table. The column's name is a misnomer (cryptographic hashing is not encryption), but is kept backward compatibility.
```
